### PR TITLE
engine: fix panic in pebble db

### LINKED ISSUE #5720: demo service panic when running pebble compaction
### What did you do?

After https://github.com/pingcap/tiflow/pull/5440, the e2e test of dataflow engine failed. ref: https://github.com/pingcap/tiflow/runs/6702976723?check_suite_focus=true

### What did you expect to see?

- If disable `withTablePropertyCollectors`, the e2e test can pass, should we add an option in `config.DBConfig`, https://github.com/pingcap/tiflow/blob/f7bf3007b1bea33ab9ad05c54f1b02493c8bda21/pkg/db/pebble.go#L117
- Besides we need to confirm whether this panic can happen in TiCDC.

### What did you see instead?

```golang
starting demo, dir /data/demo/ addr 0.0.0.0:1234
[2022/06/02 05:58:46.561 +00:00] [INFO] [kvdatamake.go:184] ["grpc serving .."]
[2022/06/02 06:02:44.561 +00:00] [INFO] [kvdatamake.go:127] ["Start to generate data ..."]
panic: runtime error: slice bounds out of range [12:1]

goroutine 85 [running]:
github.com/pingcap/tiflow/cdc/sorter/encoding.DecodeCRTs(...)
        github.com/pingcap/tiflow/cdc/sorter/encoding/key.go:42
github.com/pingcap/tiflow/pkg/db.(*tableCRTsCollector).Add(0xc00039e300?, {{0xc0001148f8?, 0x0?, 0x0?}, 0x5?}, {0xc006e80010?, 0x1a?, 0x0?})
        github.com/pingcap/tiflow/pkg/db/pebble.go:306 +0x75
github.com/cockroachdb/pebble/sstable.(*Writer).addPoint(0xc00039e300, {{0xc0001148f8?, 0x4?, 0x18bafef?}, 0x8?}, {0xc006e80010, 0x7, 0xfffff0})
        github.com/cockroachdb/pebble@v0.0.0-20211124172904-3ca75111760c/sstable/writer.go:240 +0x3a3
github.com/cockroachdb/pebble/sstable.(*Writer).Add(0xc00079e080?, {{0xc0001148f8?, 0x5a?, 0xc0001008c0?}, 0x2fec820?}, {0xc006e80010?, 0x1c18798?, 0x1c01810?})
        github.com/cockroachdb/pebble@v0.0.0-20211124172904-3ca75111760c/sstable/writer.go:219 +0x4b
github.com/cockroachdb/pebble.(*DB).runCompaction(0xc000279400, 0x4, 0xc0004d0280, {0x1c05580, 0x309a140?})
        github.com/cockroachdb/pebble@v0.0.0-20211124172904-3ca75111760c/compaction.go:2423 +0x1d75
github.com/cockroachdb/pebble.(*DB).flush1(0xc000279400)
        github.com/cockroachdb/pebble@v0.0.0-20211124172904-3ca75111760c/compaction.go:1490 +0x371
github.com/cockroachdb/pebble.(*DB).flush.func1({0x1c18f08, 0xc0007920c0})
        github.com/cockroachdb/pebble@v0.0.0-20211124172904-3ca75111760c/compaction.go:1424 +0x91
runtime/pprof.Do({0x1c18e98?, 0xc000114000?}, {{0xc0000a7ba0?, 0x0?, 0x145a7e0?}}, 0xc0005737a8)
        runtime/pprof/runtime.go:40 +0xa3
github.com/cockroachdb/pebble.(*DB).flush(0xc000573798?)
        github.com/cockroachdb/pebble@v0.0.0-20211124172904-3ca75111760c/compaction.go:1421 +0x58
created by github.com/cockroachdb/pebble.(*DB).maybeScheduleFlush
        github.com/cockroachdb/pebble@v0.0.0-20211124172904-3ca75111760c/compaction.go:1338 +0xea
```

### Versions of the cluster

Dataflow Engine version (run `df-master -V` or `df-executor -V`):

master@https://github.com/pingcap/tiflow/commit/f7bf3007b1bea33ab9ad05c54f1b02493c8bda21

### PR DESCRIPTION
**Pull request by CharlesCheung96**
### What problem does this PR solve?

Issue Number: close #5720

### What is changed and how it works?
Add a parameter to `OpenPebble` to control whether to enable `TablePropertyCollectors`

#### Tests 

 - Unit test

### Release note 

```release-note
`None`.
```

### CODE REVIEW

File: pkg/db/pebble.go [hunk 1]
```diff
@@ -102,9 +102,12 @@ func buildPebbleOption(id int, memInByte int, cfg *config.DBConfig) (pebble.Opti
...
 // OpenPebble opens a pebble.
 func OpenPebble(
-	ctx context.Context, id int, path string, memInByte int, cfg *config.DBConfig,
+	ctx context.Context,
+	id int, path string, memInByte int,
+	withTablePropertyCollectors bool,
+	cfg *config.DBConfig,
 ) (DB, error) {
-	return doOpenPebble(ctx, id, path, memInByte, true, cfg)
+	return doOpenPebble(ctx, id, path, memInByte, withTablePropertyCollectors, cfg)
```
Comment (hicqu) [inline pkg/db/pebble.go:110]: So `doOpenPebble` is not necessary. We can remove it.

File: cdc/sorter/leveldb/reader_test.go [hunk 2]
```diff
@@ -314,7 +314,8 @@ func prepareTxnData(
 	t *testing.T, r *reader, txnCount, txnSize int,
 ) db.DB {
 	cfg := config.GetDefaultServerConfig().Clone().Debug.DB
-	db, err := db.OpenPebble(context.Background(), 1, t.TempDir(), 0, cfg)
+	db, err := db.OpenPebble(context.Background(), 1, t.TempDir(), cfg,
+		db.WithTableCRTsCollectors())
```
Comment (overvenus) [inline cdc/sorter/leveldb/reader_test.go:318]:
```suggestion
	db, err := db.OpenPebble(
		context.Background(), 1, t.TempDir(), cfg, db.WithTableCRTsCollectors())
```

File: cdc/sorter/leveldb/system/system.go [hunk 3]
```diff
@@ -138,7 +138,9 @@ func (s *System) Start(ctx context.Context) error {
 	memInBytePerDB := float64(totalMemory) * s.memPercentage / float64(s.cfg.Count)
 	for id := 0; id < s.cfg.Count; id++ {
 		// Open db.
-		db, err := db.OpenPebble(ctx, id, s.dir, int(memInBytePerDB), s.cfg)
+		db, err := db.OpenPebble(ctx, id, s.dir, s.cfg,
+			db.WithCache(int(memInBytePerDB)),
+			db.WithTableCRTsCollectors())
```
Comment (overvenus) [inline cdc/sorter/leveldb/system/system.go:143]:
```suggestion
		db, err := db.OpenPebble(
			ctx, id, s.dir, s.cfg, db.WithCache(int(memInBytePerDB)), db.WithTableCRTsCollectors())
```

**Review by overvenus — APPROVED** [review]

File: pkg/db/pebble.go [hunk 4]
```diff
@@ -47,14 +47,40 @@ func (logger *pebbleLogger) Fatalf(format string, args ...interface{}) {
 	log.Panic(fmt.Sprintf(format, args...), zap.Int("db", logger.id))
 }
 
+// Option ...
```
Comment (amyangfei) [inline pkg/db/pebble.go:50]: Please complete the comment

File: pkg/db/pebble.go [hunk 5]
```diff
@@ -47,14 +47,40 @@ func (logger *pebbleLogger) Fatalf(format string, args ...interface{}) {
...
+	return func(o *pebble.Options) {
+		o.TablePropertyCollectors = append(o.TablePropertyCollectors,
+			func() pebble.TablePropertyCollector {
+				return &tableCRTsCollector{minTs: math.MaxUint64, maxTs: 0}
+			})
+	}
+}
+
 // TODO: Update DB config once we switch to pebble,
 //       as some configs are not applicable to pebble.
```
Comment (amyangfei) [inline pkg/db/pebble.go:72]: Can remove this comment now.

**Review by amyangfei — APPROVED** [review]

**Pull request merged**